reasoner/planner/PartialOrderReductionTest.java

```diff
@@ -0,0 +1,231 @@
+        {
+            RecursivePorPlanner.PartialOrderReductionSearch porSearch = new RecursivePorPlanner.PartialOrderReductionSearch(resolvables, set(), dependencies, true);
+            assertEquals(4, porSearch.allOrderings().size()); // One per start point
+        }
+        {
+            RecursivePorPlanner.PartialOrderReductionSearch porSearch = new RecursivePorPlanner.PartialOrderReductionSearch(resolvables, getVariablesByName(resolvables, set("a")), dependencies, true);
+            assertEquals(2, porSearch.allOrderings().size());
+        }
+    }
+
+    @Test
+    public void test_diamond_4_connectedness() {
+        // The relation variables need to be explicit so that they don't all wrongly become $_0
+        Resolvable<?> c1 = Concludable.create(parse("{ $r1 ($a,$b) isa r1; }")).iterator().next();
+        Resolvable<?> c2 = Concludable.create(parse("{ $r2 ($a,$c) isa r2; }")).iterator().next();
+        Resolvable<?> c3 = Concludable.create(parse("{ $r3 ($b,$d) isa r3; }")).iterator().next();
+        Resolvable<?> c4 = Concludable.create(parse("{ $r4 ($c,$d) isa r4; }")).iterator().next();
+        Set<Resolvable<?>> resolvables = set(c1, c2, c3, c4);
+
+        Map<Resolvable<?>, Set<Variable>> dependencies = new HashMap<>();
+        resolvables.forEach(resolvable -> dependencies.put(resolvable, set()));
+        {
+            RecursivePorPlanner.PartialOrderReductionSearch porSearch = new RecursivePorPlanner.PartialOrderReductionSearch(resolvables, set(), dependencies, true);
+            assertEquals(12, porSearch.allOrderings().size()); // 3 per start point
+        }
+        {
+            RecursivePorPlanner.PartialOrderReductionSearch porSearch = new RecursivePorPlanner.PartialOrderReductionSearch(resolvables, getVariablesByName(resolvables, set("a")), dependencies, true);
+            assertEquals(6, porSearch.allOrderings().size()); // 3 per start point
+        }
+        {
+            RecursivePorPlanner.PartialOrderReductionSearch porSearch = new RecursivePorPlanner.PartialOrderReductionSearch(resolvables, getVariablesByName(resolvables, set("a", "c")), dependencies, true);
+            assertEquals(10, porSearch.allOrderings().size()); // 3 per start point
+        }
+    }
+
+    private Conjunction parse(String query) {
+        return Disjunction.create(TypeQL.parsePattern(query).asConjunction().normalise()).conjunctions().iterator().next();
+    }
+
+    private Set<Variable> getVariablesByName(Set<Resolvable<?>> resolvables, Set<String> names) {
+        Set<Variable> vars = iterate(resolvables).flatMap(r -> iterate(r.variables()))
+                .filter(v -> v.id().isName() && names.contains(v.id().asName().name()))
+                .toSet();
+        assert vars.size() == names.size();
+        return vars;
+    }
+
+    private void debugPrint(List<Resolvable<?>> resolvables, List<String> names, List<List<Resolvable<?>>> allOrderings) {
```

Comment:
you should use a Logger instance, and call it something more specific like `logOrderings` rather than just "debugPrint" -- what is being printed? The printing kinda implies use for debugging as well, so no point having that in the name